src/Lean/SubExpr.lean

```diff
@@ -1,17 +1,102 @@
 /-
 Copyright (c) 2021 Microsoft Corporation. All rights reserved.
 Released under Apache 2.0 license as described in the file LICENSE.
-Authors: Sebastian Ullrich, Daniel Selsam, Wojciech Nawrocki
+Authors: Sebastian Ullrich, Daniel Selsam, Wojciech Nawrocki, E.W.Ayers
 -/
 import Lean.Meta.Basic
+import Lean.Data.Json
 import Std.Data.RBMap
 
 namespace Lean
 
 /-- A position of a subexpression in an expression.
 
 See docstring of `SubExpr` for more detail.-/
-abbrev SubExpr.Pos := Nat
+def SubExpr.Pos := Nat
+
+namespace SubExpr.Pos
+
+def maxChildren := 4
+
+/-- The coordinate `3 = maxChildren - 1` is
+reserved to denote the type of the expression. -/
+def typeCoord : Nat := maxChildren - 1
+
+def asNat : Pos → Nat := id
+
+instance : Inhabited Pos := show Inhabited Nat by infer_instance
+instance : Ord Pos := show Ord Nat by infer_instance
+instance : FromJson Pos := show FromJson Nat by infer_instance
+instance : ToJson Pos := show ToJson Nat by infer_instance
+instance : Repr Pos := show Repr Nat by infer_instance
+instance : ToString Pos := show ToString Nat by infer_instance
+
+/-- The Pos representing the root subexpression. -/
+def root : Pos := (1 : Nat)
+
+def isRoot (p : Pos) : Bool := p.asNat == 1
+
+/-- The coordinate deepest in the Pos. -/
+def head (p : Pos) : Nat :=
+  if p.isRoot then panic! "already at top"
+  else p.asNat % maxChildren
+
+def tail (p : Pos) : Pos :=
+  if p.isRoot then panic! "already at top"
+  else (p.asNat - p.head) / maxChildren
+
+def push (p : Pos) (c : Nat) : Pos :=
+  if c >= maxChildren then panic! s!"invalid coordinate {c}"
+  else p.asNat * maxChildren + c
+
+variable {α : Type} [Inhabited α]
+
+/-- Fold over the position starting at the root and heading to the leaf-/
+def foldl  (f : α → Nat → α) : α → Pos → α :=
```

Comment:
`foldup`? `folddown`? When you imagine an expression tree is the root at the top or the bottom?